topic.go

```diff
@@ -0,0 +1,115 @@
+package kafkalock
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/Shopify/sarama"
+)
+
+func validateTopic(config *Config) error {
+	if config.TopicPolicy == TopicPolicyNoValidation {
+		return nil
+	}
+
+	kafkaConfig := *config.KafkaConfig
+	kafkaConfig.Metadata.Full = true
+	admin, err := sarama.NewClusterAdmin(config.BootstrapServers, &kafkaConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create cluster admin: %w", err)
+	}
+	defer admin.Close()
+
+	topics, err := admin.ListTopics()
+	if err != nil {
+		return fmt.Errorf("failed to list topics: %w", err)
+	}
+
+	topicDetail := &sarama.TopicDetail{
+		NumPartitions:     int32(config.MaxHolders),
+		ReplicationFactor: 1,
+	}
+
+	topic, ok := topics[config.Topic]
+	if !ok {
+		if config.TopicPolicy == TopicPolicyValidateOnly {
+			return fmt.Errorf("topic \"%s\" doesn't exist", config.Topic)
+		}
+
+		config.Logger.Infow("topic doesn't exist, going to create it", "partitions", config.MaxHolders)
+
+		return createTopic(admin, config.Topic, topicDetail)
+	}
+
+	if isTopicValid(&topic, topicDetail) {
+		return nil
+	}
+
+	if config.TopicPolicy != TopicPolicyDropAndCreate {
+		message := "topic \"%s\" isn't configured properly. " +
+			"Expected number of partitions to be %d (got %d) and replication factor at least 1 (got %d)"
+
+		return fmt.Errorf(message, config.Topic, config.MaxHolders, topic.NumPartitions, topic.ReplicationFactor)
+	}
+
+	config.Logger.Infow("topic is invalid, going to re-create it",
+		"oldPartitions", topic.NumPartitions,
+		"partitions", config.MaxHolders,
+		"oldReplicationFactor", topic.ReplicationFactor,
+		"replicationFactor", 1,
+	)
+
+	err = admin.DeleteTopic(config.Topic)
+	if err != nil {
+		return fmt.Errorf("failed to delete topic: %w", err)
+	}
+
+	err = waitForTopicDeletionAndCreateTopic(admin, config.Topic, topicDetail)
+	if err != nil {
+		return fmt.Errorf(
+			"topic deletion didn't propage through the cluster before configured timeout was reached: %w",
+			err,
+		)
+	}
+
+	return nil
+}
+
+func isTopicValid(topic *sarama.TopicDetail, expectedTopic *sarama.TopicDetail) bool {
+	return topic.NumPartitions == expectedTopic.NumPartitions &&
+		topic.ReplicationFactor >= expectedTopic.ReplicationFactor
+}
+
+func createTopic(admin sarama.ClusterAdmin, name string, topic *sarama.TopicDetail) error {
+	err := admin.CreateTopic(name, topic, false)
+	if err != nil {
+		return fmt.Errorf("failed to create topic: %w", err)
+	}
+
+	return nil
+}
+
+func waitForTopicDeletionAndCreateTopic(admin sarama.ClusterAdmin, name string, topic *sarama.TopicDetail) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	for {
```

Comment:
I this while?

Reply:
There is no "while" in Go, go yeah, it is it's equivalent. In Go, there is only a single "looping" keyword and that is `for`